gazebo_ros/scripts/spawn_entity.py

```diff
@@ -0,0 +1,333 @@
+        #     help='initialize the specified joint at the specified position')
+
+        parser.add_argument('-package_to_model', action='store_true', help='convert urdf \
+                            <mesh filename="package://..." to <mesh filename="model://..."')
+
+        # TODO(shivesh): Wait for https://github.com/ros2/rclpy/issues/244
+        # parser.add_argument('-b', dest='bond', action='store_true', help='bond to gazebo \
+        #                      and delete the entity when this program is interrupted')
+        self.args = parser.parse_args(args[1:])
+
+        # TODO(shivesh): Wait for /set_model_configuration
+        # (https://github.com/ros-simulation/gazebo_ros_pkgs/issues/779)
+        # Convert position of joints to floats
+        # for i in range(len(self.args.joints)):
+        #     self.args.joints[i][1] = float(self.args.joints[i][1])
+
+    def run(self):
+        """
+        Run node, spawning entity and doing other actions as configured in program arguments.
+
+        Returns exit code, 1 for failure, 0 for success
+        """
+        # Wait for entity to exist if wait flag is enabled
+        if self.args.wait:
+            self.entity_exists = False
+
+            # TODO(shivesh): Wait for /model_states
+            # (https://github.com/ros-simulation/gazebo_ros_pkgs/issues/779)
+            # def entity_cb(entity):
+            #     self.entity_exists = self.args.wait in entity.name
+            #
+            # self.subscription = self.create_subscription(
+            #     ModelStates, '%s/model_states' % self.args.gazebo_namespace, entity_cb, 10)
+
+            self.get_logger().info(
+                'Waiting for entity {} before proceeding.'.format(self.args.wait))
+
+            while rclpy.ok() and not self.entity_exists:
+                pass
+
+        # Load entity XML from file
+        if self.args.file:
+            self.get_logger().info('Loading entity XML from file %s' % self.args.file)
+            if not os.path.exists(self.args.file):
+                self.get_logger().error('Error: specified file %s does not exist', self.args.file)
+                return 1
+            if not os.path.isfile(self.args.file):
+                self.get_logger().error('Error: specified file %s is not a file', self.args.file)
+                return 1
+            # load file
+            try:
+                f = open(self.args.file, 'r')
+                entity_xml = f.read()
+            except IOError as e:
+                self.get_logger().error('Error reading file {}: {}'.format(self.args.file, e))
+                return 1
+            if entity_xml == '':
+                self.get_logger().error('Error: file %s is empty', self.args.file)
+                return 1
+        # Load entity XML from ROS param
+        elif self.args.param:
+            self.get_logger().info('Loading entity XML from ros parameter %s' % self.args.param)
+            entity_xml = self.get_parameter(self.args.param)
```

Comment:
Even if we allow undeclared params, setting the parameter before starting the node would not work. And if we start the node first, then it crashes because the parameter is not passed yet. I don't find any use of the param. I think we can change this to a normal argument itself

Reply:
Ah I think you're right.

Getting the robot description from a param was probably the most common use case on ROS 1, because it was common to put the description on a global parameter `robot_description` that many tools used, like RViz. But since ROS2 [doesn't have](https://design.ros2.org/articles/ros_parameters.html) a central parameter server, we can't use that anymore.

It sounds like the new approach being currently used between `robot_state_publisher` and `rviz` is that the former publishes the model description in a latched topic (i.e. transient local) and the latter subscribes to it.

So how about deprecating the `-param` option and having a `-topic` one instead? Note that while the param on ROS 1 held the whole URDF file, the topic on ROS 2 holds the path to the URDF file.

* PR publishing robot description: https://github.com/ros2/robot_state_publisher/pull/9
* PR subscribing to robot description: https://github.com/ros2/rviz/pull/378

I'm also going to get the ball rolling so that this is documented on the [ROS 2 Migration guide](https://index.ros.org/doc/ros2/Contributing/Migration-Guide/).